### PR TITLE
feat: add amendment info

### PR DESCRIPTION
**Pull request by pdp2121**
## High Level Overview of Change

Add a table to store amendments information:

- Amendment’s hash

- Amendment’s name

- Enabled rippled version

This will also provide more stable source of truth for AmendmentEnabled transaction in Explorer, which hits 2 repos every time it was called.

This information will be accessed via `/network/amendments/info` endpoint.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### CODE REVIEW

File: src/api/routes/v1/amendments.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
+import { Request, Response } from 'express'
```
Comment (mvadari) [inline src/api/routes/v1/amendments.ts:1]: Is the API stuff supposed to be in a different PR?

File: package.json [hunk 2]
```diff
@@ -31,6 +31,7 @@
   "homepage": "https://github.com/ripple/validator-history-service#readme",
   "devDependencies": {
     "@types/axios": "^0.14.0",
+    "@types/create-hash": "^1.2.2",
```
Comment (mvadari) [inline package.json:34]: I assume these changes aren't supposed to be in this PR...

File: src/api/routes/v1/index.ts [hunk 3]
```diff
@@ -1,17 +1,22 @@
 import { Router as createRouter } from 'express'
 
+import { handleAmendmentsInfo, handleAmendmentInfo } from './amendments'
 import handleDailyScores from './daily-report'
 import getNetworkOrAdd from './get-network'
 import handleHealth from './health'
 import handleValidatorManifest from './manifests'
 import { handleNode, handleNodes, handleTopology } from './nodes'
 import { handleValidator, handleValidators } from './validator'
+// eslint-disable-next-line import/max-dependencies -- Disabled since this module requires multiple dependencies.
```
Comment (mvadari) [inline src/api/routes/v1/index.ts:10]: This will be easier to read if it's disabled at the top for the whole file.

File: src/api/routes/v1/info.ts [hunk 4]
```diff
@@ -56,6 +56,16 @@ const info = {
...
       example: 'https://data.xrpl.org/v1/network/validator_reports',
     },
+    {
+      action: 'Get Amendments General Information',
+      route: '/v1/network/amendments/info',
+      example: 'https://data.xrpl.org/v1/network/amendments/info',
+    },
+    {
+      action: 'Get Amendment General Information by Name or ID',
+      route: '/v1/network/amendment/info',
```
Comment (mvadari) [inline src/api/routes/v1/info.ts:66]:
```suggestion
      route: '/v1/network/amendment/info/{param}',
```

File: src/api/routes/v1/info.ts [hunk 5]
```diff
@@ -56,6 +56,16 @@ const info = {
...
+      example: 'https://data.xrpl.org/v1/network/amendment/info/{param}',
```
Comment (mvadari) [inline src/api/routes/v1/info.ts:67]:
```suggestion
      route: '/v1/network/amendment/info/{amendment}',
      example: 'https://data.xrpl.org/v1/network/amendment/info/{amendment}',
```

File: src/connection-manager/amendments.ts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+const log = logger({ name: 'amendments' })
+
+const cachedAmendmentIDs = new Map<
+  string,
+  { name: string; deprecated: boolean }
+>()
+const cachedRippledVersions = new Map<string, string>()
+
+// TODO: fix these regex linting issues later.
```
Comment (mvadari) [inline src/connection-manager/amendments.ts:16]: bump on this TODO

File: src/connection-manager/amendments.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+/**
+ * Maps the id of Amendments to its corresponding names.
+ *
+ * @returns Void.
+ */
+async function nameOfAmendmentID(): Promise<void> {
+  // The Amendment ID is the hash of the Amendment name
+  const amendmentNames = await fetchAmendmentNames()
+  if (amendmentNames !== undefined) {
```
Comment (mvadari) [inline src/connection-manager/amendments.ts:77]:
```suggestion
  if (amendmentNames != null) {
```

File: src/connection-manager/amendments.ts [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+ *
+ * @returns Void.
+ */
+async function fetchMinRippledVersions(): Promise<void> {
+  try {
+    const response = await axios.get(
+      'https://raw.githubusercontent.com/XRPLF/xrpl-dev-portal/master/content/resources/known-amendments.md',
+    )
+    const text = response.data
+    const regex =
```
Comment (mvadari) [inline src/connection-manager/amendments.ts:98]: What is this regex for? This variable should be more descriptive.

File: src/api/routes/v1/amendments.ts (hunk 1 above)
Comment (pdp2121) [reply src/api/routes/v1/amendments.ts:1]: This would just be for the `amendment/info` endpoint. The other PRs would add to validators and `amendment/vote` endpoints

File: src/connection-manager/amendments.ts (hunk 6 above)
Comment (ckniffen) [reply src/connection-manager/amendments.ts:16]: It should just require adding `/u`.

**Comment by ckniffen** [comment]
I think this code should mostly be reworked into a tool in this repo to build the initial amendment info json files.  That is because the datasources used here will likely always be behind test networks and behind when main net activates and amendment which we should catch when it happens so the data will already be there.

`setupAmendmentsInfoTable` should pull from json like the other PR does.

**Comment by pdp2121** [comment]
> I think this code should mostly be reworked into a tool in this repo to build the initial amendment info json files. That is because the datasources used here will likely always be behind test networks and behind when main net activates and amendment which we should catch when it happens so the data will already be there.
> 
> `setupAmendmentsInfoTable` should pull from json like the other PR does.

This is done only once anyway when we deploy so I don't think it's necessary to write it to json (it's complicated to write to a json file on the server as I mentioned earlier with the paths complication, reading is simple though). I have moved it to happen immediately after `setupTables` before connection began to ensure it happens before any connection, and add a log for future debug. Please let me know if that works.

File: ARCHITECTURE.md [hunk 9]
```diff
@@ -23,6 +23,8 @@ There are 3 folders in `src`, corresponding to the 3 processes that the VHS runs
   * `/network/validator/:publicKey`: Returns information about a specific validator.
   * `/network/validator/:publicKey/manifests`: Returns the manifests of a specific validator.
   * `/network/validator/:publicKey/reports`: Returns more detailed information about the reliability of a specific validator.
+  * `/network/amendments/info`: Returns general information about known amendments.
+  * `/network/amendments/info/param`: Returns general information about a specific amendment by name or ID.
```
Comment (ckniffen) [inline ARCHITECTURE.md:27]:
```suggestion
  * `/network/amendments/info/:param`: Returns general information about a specific amendment by name or ID.
```

File: src/api/routes/v1/amendments.ts [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+ * @param req - Unused express request.
+ * @param res - Express response.
+ */
+export async function handleAmendmentInfo(
+  req: Request,
+  res: Response,
+): Promise<void> {
+  try {
+    const { param } = req.params
+    if (Date.now() - cache.time > 60 * 1000) {
```
Comment (ckniffen) [inline src/api/routes/v1/amendments.ts:92]: @pdp2121 were these magic numbers fixed by another PR?

**Review by ckniffen — APPROVED** [review]

File: src/api/routes/v1/amendments.ts (hunk 10 above)
Comment (pdp2121) [reply src/api/routes/v1/amendments.ts:92]: Yes, these would be fixed in `amendment/vote` PR

File: src/connection-manager/amendments.ts [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+const cachedAmendmentIDs = new Map<
+  string,
+  { name: string; deprecated: boolean }
+>()
+const cachedRippledVersions = new Map<string, string>()
+
+const ACTIVE_AMENDMENT_REGEX =
+  /^\s*REGISTER_F[A-Z]+\s*\((?<amendmendName>\S+),\s*.*$/u
+const RETIRED_AMENDMENT_REGEX =
+  /^ .*retireFeature\("(?<amendmendName>\S+)"\)[,;].*$/u
```
Comment (mvadari) [inline src/connection-manager/amendments.ts:19]:
```suggestion
const ACTIVE_AMENDMENT_REGEX =
  /^\s*REGISTER_F[A-Z]+\s*\((?<amendmentName>\S+),\s*.*$/u
const RETIRED_AMENDMENT_REGEX =
  /^ .*retireFeature\("(?<amendmentName>\S+)"\)[,;].*$/u
```

File: src/connection-manager/amendments.ts [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+const AMENDMENT_VERSION_REGEX =
+  /\| \[(?<amendmentName>[a-zA-Z0-9_]+)\][^\n]+\| (?<version>v[0-9]*\.[0-9]*\.[0-9]*|TBD) *\|/u
+
+/**
+ * Fetch a list of amendments names from rippled file.
+ *
+ * @returns The list of amendment names.
+ */
+async function fetchAmendmentNames(): Promise<Map<string, boolean> | null> {
```
Comment (mvadari) [inline src/connection-manager/amendments.ts:29]: This can probably be cleaned up once https://github.com/XRPLF/rippled/pull/4781 is merged

File: src/connection-manager/index.ts [hunk 13]
```diff
@@ -4,11 +4,13 @@ import Crawler from '../crawler/crawl'
...
 import agreement from './agreement'
+import fetchAmendmentInfo from './amendments'
 import startConnections from './connections'
 import { doManifestJobs } from './manifests'
 
 async function start(): Promise<void> {
   await setupTables()
+  await fetchAmendmentInfo()
   // Migrate manifests from the legacy database. This will be removed once the service has collected enough manifests.
   // await migrate()
```
Comment (jonathanlei) [inline src/connection-manager/index.ts:15]: should this line be deleted?

**Review by jonathanlei — APPROVED** [review]
LGTM other than @mvadari 's typo changes in the REGEX

File: src/connection-manager/amendments.ts [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+import { AmendmentsInfo } from '../shared/types'
+import logger from '../shared/utils/logger'
+
+const log = logger({ name: 'amendments' })
+
+const cachedAmendmentIDs = new Map<
+  string,
+  { name: string; deprecated: boolean }
+>()
+const cachedRippledVersions = new Map<string, string>()
```
Comment (mvadari) [inline src/connection-manager/amendments.ts:14]: These don't look like they're actually caches?

File: src/connection-manager/amendments.ts [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+ */
+async function saveAmendmentInfo(amendment: AmendmentsInfo): Promise<void> {
+  await query('amendments_info')
+    .insert(amendment)
+    .onConflict('id')
+    .merge()
+    .catch((err) => log.error('Error Saving AmendmentInfo', err))
+}
+
+export default async function fetchAmendmentInfo(): Promise<void> {
```
Comment (mvadari) [inline src/connection-manager/amendments.ts:127]: Why is this in connection-manager and not in db setup?

File: src/shared/database/setup.ts [hunk 16]
```diff
@@ -204,6 +205,24 @@ async function setupNetworksTable(): Promise<void> {
...
+  if (!hasAmendmentsInfo) {
+    await db().schema.createTable('amendments_info', (table) => {
+      table.string('id')
+      table.string('name')
+      table.string('rippled_version')
+      table.boolean('deprecated')
+      table.primary(['id'])
+    })
+  }
+  if (!(await db().schema.hasColumn('amendments_info', 'deprecated'))) {
```
Comment (mvadari) [inline src/shared/database/setup.ts:219]: Is this line necessary? The db isn't in prod yet.

File: src/connection-manager/amendments.ts (hunk 12 above)
Comment (pdp2121) [reply src/connection-manager/amendments.ts:29]: Nice feature! I will add a TODO

File: src/connection-manager/amendments.ts (hunk 15 above)
Comment (pdp2121) [reply src/connection-manager/amendments.ts:127]: Agree, it should be in db setup since it ran once at deploy.

File: src/api/routes/v1/amendments.ts [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+    const amendments: AmendmentsInfo[] = cache.amendments.filter(
+      (amend) => amend.name === param || amend.id === param,
+    )
+    if (amendments.length === 0) {
+      res.send({ result: 'error', message: "incorrect amendment's id/name" })
+      return
+    }
+    const response: AmendmentInfoResponse = {
+      result: 'success',
+      amendment: amendments[0],
```
Comment (mvadari) [inline src/api/routes/v1/amendments.ts:104]: Should there be a check to ensure there's exactly one option here?

File: src/api/routes/v1/amendments.ts [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+    }
+    res.send(response)
+  } catch {
+    res.send({ result: 'error', message: 'internal error' })
```
Comment (mvadari) [inline src/api/routes/v1/amendments.ts:108]: There should be a log here.

File: src/api/routes/v1/index.ts [hunk 19]
```diff
@@ -1,5 +1,7 @@
+/* eslint-disable import/max-dependencies -- Disabled since this module requires multiple dependencies. */
 import { Router as createRouter } from 'express'
 
+import { handleAmendmentsInfo, handleAmendmentInfo } from './amendments'
```
Comment (mvadari) [inline src/api/routes/v1/index.ts:4]: Can these functions have slightly more differentiated names? It's hard to distinguish between the two on first glance.

File: src/shared/database/amendments.ts [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+const ACTIVE_AMENDMENT_REGEX =
+  /^\s*REGISTER_F[A-Z]+\s*\((?<amendmentName>\S+),\s*.*$/u
+const RETIRED_AMENDMENT_REGEX =
+  /^ .*retireFeature\("(?<amendmentName>\S+)"\)[,;].*$/u
+
+const AMENDMENT_VERSION_REGEX =
+  /\| \[(?<amendmentName>[a-zA-Z0-9_]+)\][^\n]+\| (?<version>v[0-9]*\.[0-9]*\.[0-9]*|TBD) *\|/u
+
+// TODO: Clean this up when this PR is merged:
+// https://github.com/XRPLF/rippled/pull/4781
```
Comment (mvadari) [inline src/shared/database/amendments.ts:23]: Please also create an issue in the repo.

Comment (pdp2121) [reply src/shared/database/amendments.ts:23]: Created an issue:
https://github.com/ripple/validator-history-service/issues/162

File: src/shared/database/amendments.ts [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+
+/**
+ * Fetches the versions when amendments are first enabled.
+ *
+ * @returns Void.
+ */
+async function fetchMinRippledVersions(): Promise<void> {
+  try {
+    const response = await axios.get(
+      'https://raw.githubusercontent.com/XRPLF/xrpl-dev-portal/master/content/resources/known-amendments.md',
```
Comment (mvadari) [inline src/shared/database/amendments.ts:95]: This is unrelated to this PR but I wonder if there's some way that this information can also be fetched from rippled. Seems difficult.

File: src/shared/types/index.ts [hunk 22]
```diff
@@ -194,6 +194,13 @@ interface DailyAgreement {
   agreement: AgreementScore
 }
 
+interface AmendmentsInfo {
+  id: string
+  name: string
+  rippled_version?: string | null
+  deprecated: boolean | null
```
Comment (mvadari) [inline src/shared/types/index.ts:201]: Why is it possible for this to be null?

File: ARCHITECTURE.md [hunk 23]
```diff
@@ -106,6 +108,30 @@ This table keeps track of the manifests of the validators.
...
+### `amendments_info`
+
+This table keeps track of the general information of all known amendments.
+
+| Key                  | Definition                                                 |
+|----------------------|------------------------------------------------------------|
+| `id`                 |The amendment id.                                           |
+| `name`               |The name of the amendment.                                  |
+| `rippled_version`    |The rippled version when the amendment is first enabled     |
+| `deprecated`         |Whether the amendment has been deprecated or retired        |
```
Comment (mvadari) [inline ARCHITECTURE.md:120]: nit:
```suggestion
| `deprecated`         |Whether the amendment has been deprecated/retired            |
```
Otherwise, it kind of reads as a boolean where the options are `deprecated` and `retired`.

File: src/shared/types/index.ts (hunk 22 above)
Comment (pdp2121) [reply src/shared/types/index.ts]: You're right, it is not possible.

File: src/shared/database/amendments.ts (hunk 21 above)
Comment (pdp2121) [reply src/shared/database/amendments.ts:95]: @intelliot @ckniffen any ideas on how we can fetch rippled version of each amendment by id?

Comment (ckniffen) [reply src/shared/database/amendments.ts:95]: The docs is the only place i know of.
